src/mgr.sol

```diff
@@ -0,0 +1,279 @@
+    function gem(bytes32,address) external returns (uint);
+    function hope(address) external;
+}
+
+interface RedeemLike {
+    function redeemOrder(uint) external;
+    function disburse(uint) external returns (uint,uint,uint,uint);
+}
+
+// This contract is essentially a merge of
+// a join and a cdp-manager.
+
+// It manages only one urn, which can be liquidated in two stages:
+// 1) In the first stage, set safe = false and call
+// pool.disburse() to try to recover as much dai as possible.
+
+// 2) After the first liquidation period has completed, we either managed to redeem
+// enough dai to wipe off all cdp debt, or this debt needs to be written off
+// and addded to the sin.
+
+// Note that the internal gem created as a result of `join` through this manager is
+// not only DROP as an ERC20 balance in this contract, but also what's currently
+// undergoing redemption from the Tinlake pool.
+
+contract TinlakeManager is LibNote {
+    // --- Auth ---
+    mapping (address => uint) public wards;
+    function rely(address usr) external note auth { require(live, "TinlakeMgr/not-live"); wards[usr] = 1; }
+    function deny(address usr) external note auth { require(live, "TinlakeMgr/not-live"); wards[usr] = 0; }
+    modifier auth {
+        require(wards[msg.sender] == 1, "TinlakeMgr/not-authorized");
+        _;
+    }
+    
+    modifier ownerOnly {
+        require(msg.sender == owner, "TinlakeMgr/owner-only");
+        _;
+    }
+
+    // The owner manages the cdp, but is not authorized to call kick or cage.
+    address public owner;
+
+    bool public safe; // Soft liquidation not triggered
+    bool public glad; // Write-off not triggered
+    bool public live; // Global settlement not triggered
+
+    uint public tab;  // Dai written off
+    bytes32 public ilk; // Constant (TODO: hardcode)
+
+    // --- Contracts ---
+    // These can all be hardcoded upon release.
+    // dss components
+    VatLike public vat;
+    GemLike public dai;
+    VowLike public vow;
+    GemJoinLike public daiJoin;
+
+    // Tinlake components
+    GemLike      public gem;
+    GemLike      public tin;
+    RedeemLike   public pool;
+
+    uint public constant dec = 18;
+
+    constructor(address vat_,      address dai_,
+                address daiJoin_,  address vow_,
+                address drop_,     address pool_,
+                address owner_,    address tranche,
+                bytes32 ilk_) public {
+
+        vat = VatLike(vat_);
+        dai = GemLike(dai_);
+        vow = VowLike(vow_);
+        daiJoin = GemJoinLike(daiJoin_);
+
+        gem = GemLike(drop_);
+        pool = RedeemLike(pool_);
+
+        ilk = ilk_;
+        wards[msg.sender] = 1;
+
+        owner = owner_;
+
+        safe = true;
+        live = true;
+        glad = true;
+
+        dai.approve(daiJoin_, uint(-1));
+        vat.hope(daiJoin_);
+        gem.approve(pool_, uint(-1));
+        gem.approve(tranche, uint(-1));
```

Comment:
Hi @godsflaw 

I think it would be possible to just approve the amount before the `redeemOrder` call in the `tell` method.

I added system tests for the liquidation scenarios here:
https://github.com/centrifuge/tinlake-maker-tests

Reply:
Calling `redeemOrder` on the pool [invokes `redeemOrder` on tranche](https://github.com/centrifuge/tinlake/blob/develop/src/lender/operator.sol#L67), which in turn makes a `transferFrom` [here](https://github.com/centrifuge/tinlake/blob/develop/src/lender/tranche.sol#L137)

Reply:
you may be correct in that the `pool` does not require approval. Will investigate